File: examples/system-agent/dtdl-validate/scripts/config-rules.js
// Sensor-config shape rules.
//
// Each rule reads a Sensor entry (devicecfg.json Sensors[] element) and
// returns either:
//   { ok: true }                 -- the sensor satisfies this rule
//   { ok: false, reason: '...' } -- the sensor violates this rule (text shown
//                                   to operators / CI logs)
//
// Rules are grouped by Parameters.MetricType. A sensor of a type with no
// registered rule set passes through unchecked. Add new (MetricType) ->
// [rule, rule, ...] entries as new validation needs land.

function isStringArray(v) {
  return Array.isArray(v) && v.every(x => typeof x === 'string');
}

function isStringOrStringArray(v) {
  return typeof v === 'string' || isStringArray(v);
}

const rules = {
  // ============================================================
  // network -- Parameters.Interface / Parameters.Interfaces
  //
  // Covers the v1.0/v1.1 resolution modes documented in
  // docs/Metrics/01_CPU_NETWORK_FIELDS.md:
  //   - Bound mode (v1.0/v1.1): single 'Interface' (string) required.
  //   - Explicit list mode (v1.1): 'Interfaces' (string array or CSV string).
  //   - Auto-detect mode (v1.1): empty value / omitted → auto-detect.
  //   - 'Interface' and 'Interfaces' MUST NOT both be set (ambiguous).
  // ============================================================
  network: [
    // Interfaces, when present, must be string OR string[].
    sensor => {
      const ifaces = sensor.Parameters?.Interfaces;
      if (ifaces === undefined) return { ok: true };
      if (isStringOrStringArray(ifaces)) return { ok: true };
      return {
        ok: false,
        reason: `Parameters.Interfaces must be string or array<string>; got ${Array.isArray(ifaces) ? 'array with non-string element' : typeof ifaces}`,
      };
    },

    // Interface (singular), when present, must be a non-empty string.
    sensor => {
      const iface = sensor.Parameters?.Interface;
      if (iface === undefined) return { ok: true };
      if (typeof iface === 'string' && iface.length > 0) return { ok: true };
      return {
        ok: false,
        reason: `Parameters.Interface must be a non-empty string; got ${typeof iface === 'string' ? '""' : typeof iface}`,
      };
    },

    // Mutual exclusivity: setting both 'Interface' and a non-empty
    // 'Interfaces' is ambiguous (runtime picks Interface, but the config
    // can't express intent clearly).
    sensor => {
      const iface = sensor.Parameters?.Interface;
      const ifaces = sensor.Parameters?.Interfaces;
      const hasIface = typeof iface === 'string' && iface.length > 0;
      const hasIfaces =
        ifaces !== undefined &&
        ifaces !== '' &&
        !(Array.isArray(ifaces) && ifaces.length === 0);
      if (hasIface && hasIfaces) {
        return {
          ok: false,
          reason: "Parameters.Interface and Parameters.Interfaces both set; runtime would pick 'Interface' but the config is ambiguous",
        };
      }
      return { ok: true };
    },
  ],

  // ============================================================
  // gpio -- Parameters.PinId / Parameters.PinIds (INTEGER-ONLY)
  //
  // Mirrors the DTDL contract declared by GpioSensorParameters in
  // docs/Metrics/05_HARDWARE_FEATURE.dtdl.json:
  //   pinId  : integer        (non-negative)
  //   pinIds : Array<integer> (non-negative; empty array = auto-detect)
  // Rules only apply when metricName == 'pinState' (isSupported takes
  // no extra parameters).
  // ============================================================
  gpio: [
    // pinId, when present, must be a non-negative integer.
    sensor => {
      if (sensor.Parameters?.MetricName !== 'pinState') return { ok: true };
      const pinId = sensor.Parameters?.PinId;
      if (pinId === undefined) return { ok: true };
      if (Number.isInteger(pinId) && pinId >= 0) return { ok: true };
      return {
        ok: false,
        reason: `Parameters.PinId must be a non-negative integer; got ${JSON.stringify(pinId)} (${typeof pinId})`,
      };
    },

    // pinIds, when present, must be an Array of non-negative integers.
    sensor => {
      if (sensor.Parameters?.MetricName !== 'pinState') return { ok: true };
      const pinIds = sensor.Parameters?.PinIds;
      if (pinIds === undefined) return { ok: true };
      if (!Array.isArray(pinIds)) {
        return {
          ok: false,
          reason: `Parameters.PinIds must be an array of non-negative integers; got ${typeof pinIds}`,
        };
      }
      for (const x of pinIds) {
        if (!Number.isInteger(x) || x < 0) {
          return {
            ok: false,
            reason: `Parameters.PinIds element ${JSON.stringify(x)} is not a non-negative integer`,
          };
        }
      }
      return { ok: true };
    },

    // pinId / pinIds mutually exclusive when both non-empty.
    sensor => {
      if (sensor.Parameters?.MetricName !== 'pinState') return { ok: true };
      const hasPinId = sensor.Parameters?.PinId !== undefined;
      const pinIds = sensor.Parameters?.PinIds;
      const hasPinIds = Array.isArray(pinIds) && pinIds.length > 0;
      if (hasPinId && hasPinIds) {
        return {
          ok: false,
          reason: "Parameters.PinId and Parameters.PinIds both set non-empty; runtime would pick 'PinId' but the config is ambiguous",
        };
      }
      return { ok: true };
    },

    // metricName=isSupported MUST NOT carry PinId / PinIds.
    sensor => {
      if (sensor.Parameters?.MetricName !== 'isSupported') return { ok: true };
      const hasPinId = sensor.Parameters?.PinId !== undefined;
      const hasPinIds = sensor.Parameters?.PinIds !== undefined;
      if (hasPinId || hasPinIds) {
        return {
          ok: false,
          reason: "metricName=isSupported sensors must not carry PinId or PinIds",
        };
      }
      return { ok: true };
    },
  ],

  // Extension points (left empty -- add rules when needed):
  //   disk:        [/* require Parameters.MountPoint */],
  //   temperature: [/* Source / Sources / MetricName tri-mode check */],
};

function validate(sensor) {
  const mt = sensor.Parameters?.MetricType;
  const ruleSet = rules[mt];
  if (!ruleSet) return { ok: true, skipped: true };
  for (const r of ruleSet) {
    const res = r(sensor);
    if (!res.ok) return res;
  }
  return { ok: true };
}

module.exports = { validate, rules };

File: examples/system-agent/dtdl-validate/scripts-dotnet/ConfigRules.cs
using System.Text.Json;

namespace Weda.Dtdl.Validator;

/// <summary>
/// Sensor-config shape rules.  Each rule reads a Sensor entry (devicecfg.json
/// Sensors[] element) and returns either <see cref="RuleResult.Ok"/> or
/// <see cref="RuleResult.Fail(string)"/>.  Rules are grouped by
/// Parameters.MetricType; a sensor whose MetricType has no registered rule
/// set passes through unchecked.
///
/// Mirror of dtdl-validate/scripts/config-rules.js -- keep the two in sync.
/// </summary>
public static class ConfigRules
{
    public readonly record struct RuleResult(bool Ok, string? Reason)
    {
        public static RuleResult OkResult { get; } = new(true, null);
        public static RuleResult Fail(string reason) => new(false, reason);
    }

    public delegate RuleResult Rule(JsonElement sensor);

    /// <summary>
    /// Rule sets keyed by MetricType. Extend by adding more keys / rules.
    /// </summary>
    public static readonly IReadOnlyDictionary<string, IReadOnlyList<Rule>> Rules =
        new Dictionary<string, IReadOnlyList<Rule>>
        {
            // ============================================================
            // network -- Parameters.Interface / Parameters.Interfaces
            // ============================================================
            ["network"] = new Rule[]
            {
                InterfacesShape,
                InterfaceShape,
                MutualExclusivity,
            },

            // ============================================================
            // gpio -- Parameters.PinId / Parameters.PinIds (INTEGER-ONLY)
            // ============================================================
            ["gpio"] = new Rule[]
            {
                PinIdShape,
                PinIdsShape,
                PinIdsMutualExclusivity,
                IsSupportedMustNotCarryPinFields,
            },

            // ============================================================
            // disk -- Parameters.MountPoint required (non-empty string)
            // ============================================================
            ["disk"] = new Rule[]
            {
                MountPointRequired,
            },

            // ============================================================
            // temperature -- Parameters.Source / Parameters.Sources
            // ============================================================
            ["temperature"] = new Rule[]
            {
                SourceShape,
                SourcesShape,
                SourceMutualExclusivity,
                V10CompatMustNotCarrySourceFields,
            },
        };

    public static RuleResult Validate(JsonElement sensor)
    {
        if (!TryGetParameter(sensor, "MetricType", out var mt) ||
            mt.ValueKind != JsonValueKind.String)
            return RuleResult.OkResult; // no MetricType -> no rules apply

        if (!Rules.TryGetValue(mt.GetString()!, out var ruleSet))
            return RuleResult.OkResult; // unknown type -> pass through

        foreach (var rule in ruleSet)
        {
            var res = rule(sensor);
            if (!res.Ok) return res;
        }
        return RuleResult.OkResult;
    }

    // ---------- rule implementations ----------

    private static RuleResult InterfacesShape(JsonElement sensor)
    {
        if (!TryGetParameter(sensor, "Interfaces", out var ifaces))
            return RuleResult.OkResult;

        if (ifaces.ValueKind == JsonValueKind.String) return RuleResult.OkResult;

        if (ifaces.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in ifaces.EnumerateArray())
                if (item.ValueKind != JsonValueKind.String)
                    return RuleResult.Fail(
                        "Parameters.Interfaces must be string or array<string>; got array with non-string element");
            return RuleResult.OkResult;
        }

        return RuleResult.Fail(
            $"Parameters.Interfaces must be string or array<string>; got {ifaces.ValueKind}");
    }

    private static RuleResult InterfaceShape(JsonElement sensor)
    {
        if (!TryGetParameter(sensor, "Interface", out var iface))
            return RuleResult.OkResult;

        if (iface.ValueKind == JsonValueKind.String &&
            !string.IsNullOrEmpty(iface.GetString()))
            return RuleResult.OkResult;

        var detail = iface.ValueKind == JsonValueKind.String ? "\"\"" : iface.ValueKind.ToString();
        return RuleResult.Fail($"Parameters.Interface must be a non-empty string; got {detail}");
    }

    private static RuleResult MutualExclusivity(JsonElement sensor)
    {
        var hasIface =
            TryGetParameter(sensor, "Interface", out var iface) &&
            iface.ValueKind == JsonValueKind.String &&
            !string.IsNullOrEmpty(iface.GetString());

        var hasIfaces = false;
        if (TryGetParameter(sensor, "Interfaces", out var ifaces))
        {
            hasIfaces = ifaces.ValueKind switch
            {
                JsonValueKind.String => !string.IsNullOrEmpty(ifaces.GetString()),
                JsonValueKind.Array  => ifaces.GetArrayLength() > 0,
                _                    => false,
            };
        }

        if (hasIface && hasIfaces)
            return RuleResult.Fail(
                "Parameters.Interface and Parameters.Interfaces both set; runtime would pick 'Interface' but the config is ambiguous");

        return RuleResult.OkResult;
    }

    // ---------- gpio rules ----------

    private static RuleResult PinIdShape(JsonElement sensor)
    {
        if (!IsMetricName(sensor, "pinState")) return RuleResult.OkResult;
        if (!TryGetParameter(sensor, "PinId", out var pinId)) return RuleResult.OkResult;

        if (TryGetNonNegativeInt(pinId, out _)) return RuleResult.OkResult;
        return RuleResult.Fail(
            $"Parameters.PinId must be a non-negative integer; got {DescribeValue(pinId)}");
    }

    private static RuleResult PinIdsShape(JsonElement sensor)
    {
        if (!IsMetricName(sensor, "pinState")) return RuleResult.OkResult;
        if (!TryGetParameter(sensor, "PinIds", out var pinIds)) return RuleResult.OkResult;

        if (pinIds.ValueKind != JsonValueKind.Array)
            return RuleResult.Fail(
                $"Parameters.PinIds must be an array of non-negative integers; got {pinIds.ValueKind}");

        foreach (var item in pinIds.EnumerateArray())
        {
            if (!TryGetNonNegativeInt(item, out _))
                return RuleResult.Fail(
                    $"Parameters.PinIds element {DescribeValue(item)} is not a non-negative integer");
        }
        return RuleResult.OkResult;
    }

    private static RuleResult PinIdsMutualExclusivity(JsonElement sensor)
    {
        if (!IsMetricName(sensor, "pinState")) return RuleResult.OkResult;

        var hasPinId = TryGetParameter(sensor, "PinId", out _);
        var hasPinIds = false;
        if (TryGetParameter(sensor, "PinIds", out var pinIds) &&
            pinIds.ValueKind == JsonValueKind.Array)
        {
            hasPinIds = pinIds.GetArrayLength() > 0;
        }

        if (hasPinId && hasPinIds)
            return RuleResult.Fail(
                "Parameters.PinId and Parameters.PinIds both set non-empty; runtime would pick 'PinId' but the config is ambiguous");

        return RuleResult.OkResult;
    }

    private static RuleResult IsSupportedMustNotCarryPinFields(JsonElement sensor)
    {
        if (!IsMetricName(sensor, "isSupported")) return RuleResult.OkResult;
        if (TryGetParameter(sensor, "PinId", out _) ||
            TryGetParameter(sensor, "PinIds", out _))
        {
            return RuleResult.Fail(
                "metricName=isSupported sensors must not carry PinId or PinIds");
        }
        return RuleResult.OkResult;
    }

    private static bool IsMetricName(JsonElement sensor, string expected)
    {
        if (!TryGetParameter(sensor, "MetricName", out var mn)) return false;
        return mn.ValueKind == JsonValueKind.String &&
               string.Equals(mn.GetString(), expected, StringComparison.Ordinal);
    }

    private static bool TryGetNonNegativeInt(JsonElement v, out long value)
    {
        value = 0;
        if (v.ValueKind != JsonValueKind.Number) return false;
        if (!v.TryGetInt64(out value)) return false;
        if (value < 0) return false;
        // Reject decimals like 4.5 (TryGetInt64 returns false anyway but be explicit).
        var raw = v.GetRawText();
        if (raw.Contains('.') || raw.Contains('e') || raw.Contains('E')) return false;
        return true;
    }

    private static string DescribeValue(JsonElement v) => v.ValueKind switch
    {
        JsonValueKind.String => $"string \"{v.GetString()}\"",
        JsonValueKind.Null   => "null",
        _                    => $"{v.ValueKind} {v.GetRawText()}",
    };

    // ---------- disk rules ----------

    private static RuleResult MountPointRequired(JsonElement sensor)
    {
        if (!TryGetParameter(sensor, "MountPoint", out var mp))
            return RuleResult.Fail("disk sensors require Parameters.MountPoint (missing)");

        if (mp.ValueKind == JsonValueKind.String && !string.IsNullOrEmpty(mp.GetString()))
            return RuleResult.OkResult;

        return RuleResult.Fail(
            $"Parameters.MountPoint must be a non-empty string; got {DescribeValue(mp)}");
    }

    // ---------- temperature rules ----------

    private static RuleResult SourceShape(JsonElement sensor)
    {
        if (!TryGetParameter(sensor, "Source", out var src)) return RuleResult.OkResult;

        if (src.ValueKind == JsonValueKind.String &&
            !string.IsNullOrEmpty(src.GetString()))
            return RuleResult.OkResult;

        return RuleResult.Fail(
            $"Parameters.Source must be a non-empty string; got {DescribeValue(src)}");
    }

    private static RuleResult SourcesShape(JsonElement sensor)
    {
        if (!TryGetParameter(sensor, "Sources", out var srcs)) return RuleResult.OkResult;

        if (srcs.ValueKind != JsonValueKind.Array)
            return RuleResult.Fail(
                $"Parameters.Sources must be an array of non-empty strings; got {srcs.ValueKind}");

        foreach (var item in srcs.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.String ||
                string.IsNullOrEmpty(item.GetString()))
                return RuleResult.Fail(
                    $"Parameters.Sources element {DescribeValue(item)} is not a non-empty string");
        }
        return RuleResult.OkResult;
    }

    private static RuleResult SourceMutualExclusivity(JsonElement sensor)
    {
        var hasSrc =
            TryGetParameter(sensor, "Source", out var src) &&
            src.ValueKind == JsonValueKind.String &&
            !string.IsNullOrEmpty(src.GetString());

        var hasSrcs = false;
        if (TryGetParameter(sensor, "Sources", out var srcs) &&
            srcs.ValueKind == JsonValueKind.Array)
        {
            hasSrcs = srcs.GetArrayLength() > 0;
        }

        if (hasSrc && hasSrcs)
            return RuleResult.Fail(
                "Parameters.Source and Parameters.Sources both set non-empty; runtime would pick 'Source' but the config is ambiguous");

        return RuleResult.OkResult;
    }

    private static RuleResult V10CompatMustNotCarrySourceFields(JsonElement sensor)
    {
        if (IsMetricName(sensor, "therm")) return RuleResult.OkResult;

        var hasSrc = TryGetParameter(sensor, "Source", out _);
        var hasSrcs = TryGetParameter(sensor, "Sources", out _);
        if (!hasSrc && !hasSrcs) return RuleResult.OkResult;

        string mnDescription = "<missing>";
        if (TryGetParameter(sensor, "MetricName", out var mn) &&
            mn.ValueKind == JsonValueKind.String)
        {
            mnDescription = $"\"{mn.GetString()}\"";
        }
        return RuleResult.Fail(
            $"v1.0-compat temperature sensors (MetricName={mnDescription}) must not carry Source or Sources; the source name is the MetricName itself");
    }

    // ---------- helpers ----------

    private static bool TryGetParameter(JsonElement sensor, string key, out JsonElement value)
    {
        value = default;
        if (sensor.ValueKind != JsonValueKind.Object) return false;
        if (!sensor.TryGetProperty("Parameters", out var parameters)) return false;
        if (parameters.ValueKind != JsonValueKind.Object) return false;
        return parameters.TryGetProperty(key, out value);
    }
}
